### PR TITLE
cmake: use check_symbol_exists to check dlopen + need for dl library

### LINKED ISSUE #6283: cmake doesn't overwrite HAVE_DLOPEN variable
(Noticed this while messing with https://github.com/libsdl-org/sdl12-compat/issues/222)

We have this check for dlopen() in our cmake'ry:
https://github.com/libsdl-org/SDL/blob/main/cmake/sdlchecks.cmake#L32-L55
```
macro(CheckDLOPEN)
  check_symbol_exists(dlopen "dlfcn.h" HAVE_DLOPEN)
  if(NOT HAVE_DLOPEN)
    check_library_exists(dl dlopen "" DLOPEN_LIB)
    if(DLOPEN_LIB)
      list(APPEND EXTRA_LIBS dl)
      set(_DLLIB dl)
      set(HAVE_DLOPEN TRUE)
    endif()
  endif()
  if(HAVE_DLOPEN)
    if(_DLLIB)
      set(CMAKE_REQUIRED_LIBRARIES ${_DLLIB})
    endif()
    check_c_source_compiles("
       #include <dlfcn.h>
       int main(int argc, char **argv) {
         void *handle = dlopen(\"\", RTLD_NOW);
         const char *loaderror = (char *) dlerror();
         return 0;
       }" HAVE_DLOPEN)
    set(CMAKE_REQUIRED_LIBRARIES)
  endif()
endmacro()
```

The last check_c_source_compiles() check attempts to make sure that we
have dlfcn.h and can link, however it doesn't overwrite the HAVE_DLOPEN
variable: Verified this by changing `dlopen` to a bogus `dlopen1`. How
can we rectify this? Use another unique dummy variable?

There may be more similar cases like this in our cmake'ry. @madebr: All
yours.

### PR DESCRIPTION
**Pull request by madebr**
Check for `dlopen` in 2 stages:
1. Check whether including `dlfcn.h` is sufficient for providing a `dlopen` symbol
2. If that is not enough, check whether we need to link to `dl` as well. If so, add `dl` to the link library list.

If either one of above succeeds, then set `HAVE_DLOPEN` to a true-ish value.

Fixes #6283

### CODE REVIEW

File: cmake/sdlchecks.cmake [hunk 1]
```diff
@@ -30,28 +30,19 @@ macro(FindLibraryAndSONAME _LIB)
 endmacro()
 
 macro(CheckDLOPEN)
-  check_symbol_exists(dlopen "dlfcn.h" HAVE_DLOPEN)
-  if(NOT HAVE_DLOPEN)
-    check_library_exists(dl dlopen "" DLOPEN_LIB)
-    if(DLOPEN_LIB)
+  cmake_push_check_state(RESET)
+  check_symbol_exists(dlopen "dlfcn.h" HAVE_DLOPEN_IN_DLFCN_H_1)
```
Comment (sezero) [inline cmake/sdlchecks.cmake:34]: `HAVE_DLOPEN_IN_LIBC` looks better for this

File: cmake/sdlchecks.cmake [hunk 2]
```diff
@@ -30,28 +30,19 @@ macro(FindLibraryAndSONAME _LIB)
...
+  if(NOT HAVE_DLOPEN_IN_DLFCN_H_1)
+    set(CMAKE_REQUIRED_LIBRARIES dl)
+    check_symbol_exists(dlopen "dlfcn.h" HAVE_DLOPEN_IN_DLFCN_H_2)
```
Comment (sezero) [inline cmake/sdlchecks.cmake:37]: `HAVE_DLOPEN_IN_LIBDL` looks better for this

File: cmake/sdlchecks.cmake [hunk 3]
```diff
@@ -30,28 +30,19 @@ macro(FindLibraryAndSONAME _LIB)
...
+    if(HAVE_DLOPEN_IN_DLFCN_H)
```
Comment (sezero) [inline cmake/sdlchecks.cmake:38]: This needs replacing: your version is `HAVE_DLOPEN_IN_DLFCN_H_2`, mine's as above

File: cmake/sdlchecks.cmake [hunk 4]
```diff
@@ -30,28 +30,19 @@ macro(FindLibraryAndSONAME _LIB)
...
-    check_c_source_compiles("
-       #include <dlfcn.h>
-       int main(int argc, char **argv) {
-         void *handle = dlopen(\"\", RTLD_NOW);
-         const char *loaderror = (char *) dlerror();
-         return 0;
-       }" HAVE_DLOPEN)
-    set(CMAKE_REQUIRED_LIBRARIES)
+  if(HAVE_DLOPEN_IN_DLFCN_H_1 OR HAVE_DLOPEN_IN_DLFCN_H_2)
+    set(HAVE_DLOPEN ON)
```
Comment (sezero) [inline cmake/sdlchecks.cmake:43]: Isn't `TRUE` the better cosmetics here? (They amount to same, but..)

**Comment by sezero** [comment]
Is this to fix https://github.com/libsdl-org/SDL/issues/6283 ?

File: cmake/sdlchecks.cmake (hunk 1 above)
Comment (madebr) [reply cmake/sdlchecks.cmake]: Your suggestions sounded way better, so I applied them.

File: cmake/sdlchecks.cmake (hunk 4 above)
Comment (madebr) [reply cmake/sdlchecks.cmake]: Probably. `HAVE_DLOPEN` is aimed for `SDL_config.h` and not for some cmake dependency.

**Comment by madebr** [comment]
> Is this to fix #6283 ?

Yes, I added it to my initial message as well.

**Comment by sezero** [comment]
>   I added it to my initial message as well.

sorry, missed that

**Comment by madebr** [comment]
> > I added it to my initial message as well.
> 
> sorry, missed that

You missed nothing, I just added it :wink:

**Comment by sezero** [comment]
About #6283: Are there any other places in our cmake'ry with the same fail-to-overwrite-variable issue?

**Comment by madebr** [comment]
> About #6283: Are there any other places in our cmake'ry with the same fail-to-overwrite-variable issue?

Yes. I immediately found a potential issue in `CheckOSS`.
I will do a sweep though `sdlchecks.cmake`.

**Comment by sezero** [comment]
> Yes. I immediately found a potential issue in `CheckOSS`. I will do a sweep though `sdlchecks.cmake`.

The issue tend to happen with an already-set variable, like:
```
check_c_source_compiles("some code" HAVE_XXX)
if(HAVE_XXX)
  check_c_source_compiles("more code" HAVE_XXX) # <-- if HAVE_XXX is set TRUE, it won't be set here
endif()
```

**Comment by madebr** [comment]
> > Yes. I immediately found a potential issue in `CheckOSS`. I will do a sweep though `sdlchecks.cmake`.
> 
> The issue tend to happen with an already-set variable, like:
> 
> ```
> check_c_source_compiles("some code" HAVE_XXX)
> if(HAVE_XXX)
>   check_c_source_compiles("more code" HAVE_XXX) # <-- if HAVE_XXX is set TRUE, it won't be set here
> endif()
> ```

Indeed, I looked at the source of [`CheckSymbolExists`](https://github.com/Kitware/CMake/blob/eb6e2ef7f61de1dd4ff6ec39e886d5fa8cf9aea7/Modules/CheckSymbolExists.cmake#L94) and it does not re-run if the variable is defined.
This caching makes running cmake a 2nd time fast.
CMake does not know how and why a variable (e.g. `HAVE_DLOPEN`) was defined.
If it would retest every variable that is set to False, reconfigures would go slow(er).

**Comment by sezero** [comment]
OK, I see.

Is this ready? Does CMakeLists.txt need scanning through, too?

**Comment by madebr** [comment]
I found another offender for `clock_gettime`.
I think all other `check_*` calls use unique variable names.

**Comment by sezero** [comment]
OK, will merging.

Then let's copy the dlopen solution to sdl12-compat for https://github.com/libsdl-org/sdl12-compat/issues/222
(I have a working solution there, but I don't speak cmake :)

**Pull request merged**